runtime/sema/checker.go

```diff
@@ -2069,37 +2069,74 @@ func (checker *Checker) checkInvalidInterfaceAsType(ty Type, pos ast.HasPosition
 }
 
 func (checker *Checker) checkParameterizedTypeIsInstantiated(ty Type, pos ast.HasPosition) {
-	parameterizedType, ok := ty.(ParameterizedType)
-	if !ok {
-		return
-	}
+	switch t := ty.(type) {
```

Comment:
It might make sense to make this a function on `sema.Type` instead, so the implementation is ensured to be exhaustive for all types now and in the future.

Reply:
I didn't get you actually. Can you please share a reference if you have?

Reply:
Basically:
- Add a function `CheckInstantiated(pos ast.HasPosition, report func(err error))` to `sema.Type`
- Refactor each `case` in the switch into an implementation of `CheckInstantiated` for that `Type`, e.g. for `OptionalType` (pseudo-code, not tested):

    ```cadence
    func (t *OptionalType) CheckInstantiated(pos ast.HasPosition, report func(err error)) {
    	t.Type.CheckInstantiated(pos, report)
    }
    ```

- Inside of the function, use the `report` parameter/function instead of `checker.report`
- Change `checkParameterizedTypeIsInstantiated` to simply `ty.CheckInstantiated(pos, checker.report)`

For many types, e.g. `Void`, `String`, etc. the implementation is simply a NO-OP.

Reply:
Done, thanks for the detailed explanation!